File: lib/hecks/bluebook/behaviour/read_model.rb
module Hecks
  module Bluebook
    module Behaviour
      # **What a read model does**. Its declared half is the gathered heads
      # and the query shape; these are readings taken off them.
      module ReadModel
        # Lists the fields this read model groups rows by.
        #
        # @return [Array<Symbol>] each declared `group_by` field's name, in declaration order
        def group_by_fields = @group_by.map { |row| row[:field].to_sym }

        # `!!` rather than a bare `@count` — the DSL/reconstruction
        # boundary (ReadModel#initialize) already normalises to
        # `true`/`nil`, so this is belt and braces against a future
        # caller constructing a ReadModel by hand with `count: false`.
        #
        # @return [Boolean] whether this read model reduces to a row count
        def count? = !!@count

        # Names the verb `Dispatcher#query` looks this read model up by.
        #
        # @return [String] this read model's name in `snake_case`
        def query_name = Naming.snake(@name)

        # Which gathered heads the filtering applies to (ADR 0055) — plural,
        # since `where`/`order_by`/`limit`/`offset` can now each independently
        # name a many-side head via `on:` once there's more than one. A read
        # model with a single many-side head keeps the old reading: every
        # untargeted option (plus `group_by`/`count`/`median`, still
        # single-head-only — ADR 0055) applies to it, same as before `on:`
        # existed. With several many-side heads, only the ones actually named
        # by a targeted option are eligible.
        #
        # @return [Array<Symbol>] the `:as` name of each many-side head that filtering
        #   applies to; `[]` if this read model has no many-side head
        def filtered_head_names
          many = @aggregate_heads.select { |head| head[:many] }
          return [] if many.empty?

          return single_filtered_head_name(many) if many.one?

          targets = (wheres.map(&:target) + [order_by&.target, limit&.target, offset&.target]).compact.uniq
          targets.filter_map { |target| many.find { |head| head[:aggregate] == target.to_s } }.map { |head| head[:as] }
        end

        # The pre-`on:` reading (ADR 0055), unchanged: with exactly one
        # many-side head, every untargeted option (plus `group_by`/`count`/
        # `median`, still single-head-only) applies to it — split out only
        # to keep `filtered_head_names` itself under this file's own
        # complexity budget, not because the two questions differ in kind.
        #
        # @param many [Array<Hash{Symbol => Object}>] the read model's many-side
        #   `aggregate_heads` rows; must hold exactly one
        # @return [Array<Symbol>] `[the one head's :as name]` if any filtering option is
        #   declared, else `[]`
        def single_filtered_head_name(many)
          declared = wheres.any? || order_by || limit || offset || authorization&.tenant ||
                     @group_by.any? || count? || @median_field
          declared ? [many.first[:as]] : []
        end

        # The where/order_by/limit/offset that apply to one eligible head
        # (ADR 0055) — a small view `Ports::Query::InMemory.execute` reads
        # exactly the way it already reads a whole `Query`/`ReadModel`
        # (`.wheres`/`.order_by`/`.limit`/`.offset`/`.null_semantics`), scoped
        # to `head_as`'s own aggregate: an untargeted option applies when
        # `head_as` is the read model's one many-side head (the pre-`on:`
        # reading, unchanged) ; a targeted one applies when its `target`
        # resolves to `head_as`'s own aggregate.
        FilteredOptions = Struct.new(:wheres, :order_by, :limit, :offset, :null_semantics)

        # Scopes this read model's filtering options down to one included head.
        #
        # @param head_as [Symbol] the `:as` name of the head to scope filtering to
        # @return [FilteredOptions] the `wheres`/`order_by`/`limit`/`offset` that apply to
        #   `head_as`, and this read model's own `null_semantics`
        def options_for(head_as)
          many = @aggregate_heads.select { |head| head[:many] }
          aggregate_name = @aggregate_heads.find { |head| head[:as] == head_as }&.fetch(:aggregate)
          applies = lambda do |target|
            target.nil? ? many.one? : target.to_s == aggregate_name
          end

          FilteredOptions.new(
            wheres.select { |where| applies.call(where.target) },
            order_by && applies.call(order_by.target) ? order_by : nil,
            limit && applies.call(limit.target) ? limit : nil,
            offset && applies.call(offset.target) ? offset : nil,
            null_semantics
          )
        end
      end
    end
  end
end

File: rust/parser/src/parse/read_model.rs
//! The `ReadModel` construct (`lib/hecks/bluebook/ir/read_model.rb`,
//! `read_model` the current spelling — ADR 0025 reverts `report`, `was:
//! "report"`, still answered for frozen era text under the legacy
//! grammar). STAGE 3: `description`/`include`/`group_by` —
//! console_settings.bluebook's own `Styles`/`Curated` read models, both
//! ROOTLESS (no `reference_to`) with a single many-side `include` and a
//! `group_by` naming that same head's own fields. STAGE 4 adds
//! `reference_to` (the single-root form: `CustomerPortfolio`/
//! `ComplianceDashboard`) and the five open-map option words PLUS
//! `where`/`order_by`/`limit` — `ReadModelBuilder` `include
//! QuerySpecification::Common::DSL`, the SAME module `parse::query`
//! shares via `build::query_options`; confirmed real by
//! `ComplianceDashboard`'s own filtered/ordered/capped shape
//! (banking.bluebook's own comment names it "the one shape
//! seal_query_options allows").

use crate::build::{naming, query_derive, query_options, read_model as build_read_model};
use crate::diag::{Diagnostic, ParseResult};
use crate::ir;
use crate::lex::SourceLine;

pub fn not_implemented(file: &str, line: usize, word: &str) -> Diagnostic {
    Diagnostic::not_yet_implemented(file, line, format!("ReadModel.{word}"))
}

const OPTION_WORDS: &[&str] = &["offset", "cursor", "authorize", "nulls", "inspect_query"];

/// Parses a `report "Name" do ... end` body (`report`/`read_model`).
/// `include`s are gathered raw and resolved into `aggregate_heads` at the
/// END (mirroring `ReadModelBuilder#build`'s own build-time resolution —
/// "order-independent," per that builder's own comment), the same
/// deferred-resolution shape `parse::aggregate`'s own `identified_by`
/// handling already uses for the identical Ruby-side reason — this is
/// also why `reference_to` (needed to compute each head's own `many:`)
/// is read into a plain local rather than applied eagerly.
pub fn parse_body(
    file: &str,
    lines: &[SourceLine],
    pos: &mut usize,
    name: &str,
) -> ParseResult<ir::ReadModel> {
    let mut description: Option<String> = None;
    let mut reference_target: Option<String> = None;
    let mut reference_name: Option<String> = None;
    let mut includes: Vec<(String, Option<String>)> = Vec::new();
    let mut group_by_fields: Vec<String> = Vec::new();
    let mut count = false;
    let mut median_field: Option<String> = None;
    let mut wheres: Vec<ir::WhereClause> = Vec::new();
    let mut order_by: Option<ir::OrderBy> = None;
    let mut limit: Option<ir::LimitSpec> = None;
    let mut options = ir::QueryOptions::default();
    let mut last_line = 0usize;

    loop {
        let Some(gated) = super::next_line(file, lines, pos, "ReadModel")? else {
            break;
        };
        last_line = gated.line.number;

        match gated.row.word {
            "description" => {
                description = Some(super::positional_text(
                    file,
                    last_line,
                    "description",
                    &gated.args,
                    1,
                )?)
            }
            // `ReadModelBuilder#reference_to(type, as: nil)` — the
            // single-root form (`CustomerPortfolio`'s own `reference_to
            // Customer`). Ruby refuses a SECOND call ("already has a
            // projection reference") — a `seal`-style declaration-time
            // check, out of scope here the same way `AggregateBuilder`'s
            // own seals are (this parser's job is shape, not semantics);
            // a bluebook that violates it never produced the oracle this
            // parser is compared against in the first place.
            "reference_to" => {
                let target_raw =
                    super::positional_constant(file, last_line, "reference_to", &gated.args, 1)?;
                let target = naming::demodulise(target_raw);
                // `@reference_name = (as || Naming.snake(@reference_target)).to_sym`
                // — ALWAYS set once `reference_to` is called, defaulting
                // to the target's own snake-cased name.
                reference_name = Some(
                    super::named_symbol(&gated.args, "as")
                        .unwrap_or_else(|| naming::snake(&target)),
                );
                reference_target = Some(target);
            }
            "include" => {
                let target =
                    super::positional_constant(file, last_line, "include", &gated.args, 1)?;
                let as_name = super::named_symbol(&gated.args, "as");
                includes.push((naming::demodulise(target), as_name));
            }
            "group_by" => {
                // VARIADIC — `argument_gate`'s own `variadic: "true"`
                // handling (syntax.bluebook's own comment on that
                // column) already confirmed every positional here reads
                // as a symbol; just strip each one's leading `:`.
                group_by_fields = gated
                    .args
                    .positional
                    .iter()
                    .map(|(_, text)| text.trim().trim_start_matches(':').to_string())
                    .collect();
            }
            // `ReadModelBuilder#count` — a bare word, no argument at
            // all (`def count = @count = true`); its presence in the
            // source IS the value, same as `generic`'s own Bluebook-
            // context row.
            "count" => count = true,
            // `ReadModelBuilder#median(field)` — one required
            // positional symbol (`def median(field) = @median_field =
            // field.to_sym`).
            "median" => {
                median_field = Some(super::positional_symbol(
                    file,
                    last_line,
                    "median",
                    &gated.args,
                    1,
                )?)
            }
            "where" => {
                query_derive::refuse_on_target(file, last_line, "where", &gated.args.named)?;
                wheres.extend(query_derive::where_clauses(&gated.args.named))
            }
            "order_by" => {
                // NOT `refuse_on_target` here — unlike `where` (below),
                // `order_by` has a fixed, declared argument schema, so
                // the argument gate itself already refuses an
                // undeclared `on:` upstream, before this arm ever runs
                // (confirmed live: "'order_by' takes no 'on:'
                // argument"). Only `where`'s own open-ended `pairs`
                // shape needs a defensive check — there is no fixed
                // schema for it to be caught against.
                let field = super::positional_symbol(file, last_line, "order_by", &gated.args, 1)?;
                let direction = match gated.args.positional.iter().find(|(idx, _)| *idx == 2) {
                    Some((_, text)) => text.trim().trim_start_matches(':').to_string(),
                    None => "asc".to_string(),
                };
                order_by = Some(ir::OrderBy { field, direction });
            }
            "limit" => {
                let raw = super::positional_constant(file, last_line, "limit", &gated.args, 1)?;
                limit = Some(ir::LimitSpec {
                    value: crate::ruby_value::render(&crate::ruby_value::read(raw)),
                });
            }
            word if OPTION_WORDS.contains(&word) => {
                query_options::apply(file, last_line, word, &gated.args, &mut options)?
            }
            _ => {
                return Err(super::not_built_yet(
                    "ReadModel",
                    gated.row,
                    file,
                    last_line,
                    &gated.call.word,
                ))
            }
        }
    }

    if includes.is_empty() && reference_target.is_none() {
        return Err(Diagnostic::new(
            file,
            last_line,
            format!("{name} needs an aggregate-head reference or at least one include"),
        ));
    }

    let aggregate_heads = build_read_model::aggregate_heads(
        file,
        last_line,
        name,
        &includes,
        reference_target.as_deref(),
    )?;

    Ok(ir::ReadModel {
        name: name.to_string(),
        description,
        reference_name,
        reference_target,
        query_name: naming::snake(name),
        wheres,
        order_by,
        limit,
        aggregate_heads,
        group_by: group_by_fields,
        count,
        median_field,
        options,
    })
}
